public/app/features/explore/QueryRows.tsx

```diff
@@ -48,11 +50,26 @@ export const QueryRows = ({ exploreId }: Props) => {
     dispatch(runQueries({ exploreId }));
   }, [dispatch, exploreId]);
 
+  const sortQueryOutlineItems = useCallback(() => {
+    if (ctx) {
+      const newOutlineItems = [...ctx.outlineItems];
+      const queriesItem = newOutlineItems.find((item) => item.panelId === 'Queries');
+      if (queriesItem && queriesItem.children) {
+        queriesItem.children.sort(sortElementsByDocumentPosition);
+        newOutlineItems[newOutlineItems.findIndex((item) => item.panelId === 'Queries')] = queriesItem;
+      }
+
+      ctx.updateOutlineItems(newOutlineItems);
+    }
+  }, [ctx]);
+
   const onChange = useCallback(
-    (newQueries: DataQuery[]) => {
-      dispatch(changeQueries({ exploreId, queries: newQueries }));
+    async (newQueries: DataQuery[]) => {
+      await dispatch(changeQueries({ exploreId, queries: newQueries }));
+      // Used to sort the order in Content Outline when changing the order of query rows in the UI through drag and drop
+      sortQueryOutlineItems();
```

Comment:
We could put this in Explore component too. `QueryRows` are specific to Explore so I thought it made more sense to put it in here and make it more local. Also, Explore is a class component and can't use `useContentOutlineContext`, but I can add a consumer if we think it's better to put this logic in Explore

Reply:
After doing some testing, this is introducing a bug where if you update the title of the query row header that row will disappear from content outline. Working on fixing this

Reply:
The logic for sorting looks like quite specific to ContentOutline itself, and might be nice to reduce how much QueryRows actually know about internals of the content outline. Would it be more readable if we simply had `ctx.refresh()` method that triggers resorting all elements (shouldn't be expensive, there's not many of them)? 

In theory such `ctx.refresh()` could happen internally inside the Content Outline (watching on DOM changes with MutationObserver).

Reply:
That's a great idea! I was looking into MutationObserver, and I think that would be the best option since, as you said, there are not many elements in the content outline.

Reply:
@ifrost It looks like MutationObserver triggers a lot more updates and after adding 4 or 5 queries the updates slow down updating the content outline. The observer just focuses on the queries content outline item and filters only childList mutation types, but still triggers several thousands of items. Maybe we can look at this together and see if we can optimize observer. Another option could be to pass queries from Explore to context provider which is already in Explore and then inside provider we can just use add `refresh()` that triggers when queries change